craft_parts/utils/process.py

```diff
@@ -0,0 +1,242 @@
+import subprocess
+import sys
+from collections.abc import Generator, Sequence
+from contextlib import closing, contextmanager
+from dataclasses import dataclass
+from io import BytesIO
+from pathlib import Path
+from typing import IO, TextIO, cast
+
+Command = str | Path | Sequence[str | Path]
+Stream = IO | int | None
+
+_BUF_SIZE = 4096
+
+# Compatibility with subprocess.DEVNULL
+DEVNULL = subprocess.DEVNULL
+
+
+@dataclass
+class ProcessResult:
+    """Describes the outcome of a process."""
+
+    returncode: int
+    stdout: bytes
+    stderr: bytes
+    combined: bytes
+    command: Command
+
+    def check_returncode(self) -> None:
+        """Raise a ProcessError if the process returned non-zero."""
+        if self.returncode != 0:
+            raise ProcessError(self)
+
+
+class _ProcessStream:
+    def __init__(
+        self, read_fd: IO[bytes] | None, write_fd: IO[bytes] | IO[str]
+    ) -> None:
+        self.read_fd = read_fd
+        self.write_fd = write_fd
+
+        self._linebuf = BytesIO()
+        self._streambuf = BytesIO()
+
+        # Don't process anything if DEVNULL is the read end of the pipe
+        if read_fd:
+            self.process = self._process_something
+        else:
+            self.process = self._process_blank
+
+        # True if stream only accepts strings
+        if hasattr(write_fd, "encoding"):
+            self._write = self._write_str
+        else:
+            self._write = self._write_bytes
+
+    def __del__(self) -> None:
+        self._linebuf.close()
+        self._streambuf.close()
+
+    @property
+    def singular(self) -> bytes:
+        return self._streambuf.getvalue()
+
+    def _process_blank(self) -> bytes:
+        """Do nothing."""
+        return b""
+
+    def _process_something(self) -> bytes:
+        """Forward any data from ``self.read_fd`` to ``self.write_fd`` and return a copy of it."""
+        data = cast(IO[bytes], self.read_fd).read(_BUF_SIZE)
+        i = data.rfind(b"\n")
+        if i >= 0:
+            self._linebuf.write(data[: i + 1])
+            self._streambuf.write(self._linebuf.getvalue())
+            self._write()
+            self._linebuf.seek(0)
+            self._linebuf.write(data[i + 1 :])
+            return data
+
+        self._linebuf.write(data)
+        return b""
+
+    def _write_bytes(self) -> None:
+        cast(IO[bytes], self.write_fd).write(self._linebuf.getvalue())
+
+    def _write_str(self) -> None:
+        bytestr = self._linebuf.getvalue()
+        cast(IO[str], self.write_fd).write(str(bytestr, "utf-8", "replace"))
+
+
+def run(
+    command: Command,
+    *,
+    cwd: Path | None = None,
+    stdout: Stream = None,
+    stderr: Stream = None,
+    check: bool = True,
+) -> ProcessResult:
+    """Execute a subprocess and collect its output.
+
+    This function collects the stdout and stderr streams as separate
+    accounts and a singular, combined account.
+
+    :param command: Command to execute.
+    :type Command:
```

Comment:
pedantic: IDE/LSPs can infer the type from the arguments themselves, so you don't need to add them to the docstring.  This becomes a problem because it is easy to update the code and forget to update the docstring type